### PR TITLE
Reexport @theme-ui/components from theme-ui

### PR DESCRIPTION
**Pull request by hasparus**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://theme-ui.com/migrating/, https://github.com/system-ui/theme-ui/blob/master/packages/theme-ui/src/index.js
- [x] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.
- [x] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`. If for reason the any rule need to be disabled, disable it for that line using `// tslint:disable-next-line [ruleName]` and not for whole package so that the need for disabling can be reviewed.

**Linked to** https://github.com/system-ui/theme-ui/issues/121.

### CODE REVIEW

**Comment by prateekkathal** [comment]
@hasparus I don't think it is a good idea to introduce breaking changes. Maybe we should keep the old keys/properties to support older versions. Please add the below as a comment.

```js
DEPRACATED: Not supported after theme-ui:^0.3.*
```

Also, I'd suggest adding tests for the newer versions separately rather than changing the old ones.

**Comment by hasparus** [comment]
> Maybe we should keep the old keys/properties to support older versions. Please add the below as a comment.

I get the point. It leads to another problem though.
Comments appear in tooltips and that's just it. No errors. So we can update theme-ui to 0.3 without seeing the changelog. Maybe dependabot or greenkeeper made the PR.
We merge to PR, continous deployment runs and our default color mode changed, because **`initialColorMode` is not handled by theme-ui anymore**. This is a bug at runtime.

`@types` packages are versioned. I believe we should match breaking changes of theme-ui (best effort) to help users avoid bugs in runtime. This PR won't break old versions of @types/theme-ui.

I'd find deprecation comment fitting, if theme-ui added a deprecation warning at runtime. There is no warning, thus, this is a breaking change, and we can either help the user deal with it by removing `initialColorMode` or don't.

Connected issues: https://github.com/DefinitelyTyped/DefinitelyTyped/issues/7719 https://github.com/DefinitelyTyped/DefinitelyTyped/issues/25677

**Comment by hasparus** [comment]
We needed to upgrade required version to 3.5 because @types/theme-ui__components explodes on TS3.4 (due to "as" prop).

File: types/theme-ui/index.d.ts [hunk 1]
```diff
@@ -1,29 +1,26 @@
...
 // TypeScript Version: 3.1
 
-import { ResponsiveStyleValue, SystemStyleObject } from '@styled-system/css';
+import { SystemStyleObject } from '@styled-system/css';
 import * as CSS from 'csstype';
 import * as React from 'react';
-import { lineHeight, Theme as StyledSystemTheme } from 'styled-system';
+import { Theme as StyledSystemTheme } from 'styled-system';
 import { Interpolation, SerializedStyles } from '@emotion/serialize';
+export * from '@theme-ui/components';
```
Comment (prateekkathal) [inline types/theme-ui/index.d.ts]: I am not sure why this belongs here. The goal is to only make types exportable, not components by itself.

File: types/theme-ui/index.d.ts [hunk 2]
```diff
@@ -157,14 +154,6 @@ export interface SxProps {
 
 type SxComponent<T extends SxProps = IntrinsicSxElements['div']> = React.ComponentClass<T & { as?: React.ElementType }>;
 
-export const Box: SxComponent;
```
Comment (prateekkathal) [inline types/theme-ui/index.d.ts]: Again, the goal is to define component props, not make them exportable from the main package itself.

**Review by prateekkathal — CHANGES_REQUESTED** [review]

**Comment by prateekkathal** [comment]
@hasparus It seems like there is a PR sitting for `styled-system` #41903 & issue still unresolved for `theme-ui` #40897 

I'd defer merging this PR as the proposed changes might fix older versions of theme-ui types. If a breaking change must be introduced, it should be made ensuring previous builds don't have errors.

File: types/theme-ui/index.d.ts (hunk 1 above)
Comment (hasparus) [reply types/theme-ui/index.d.ts]: Take a look at https://github.com/system-ui/theme-ui/blob/b0257f0aa3dda6e65482552780fda715783205c3/packages/theme-ui/src/index.js#L18

This is the **title of my pull request**.
ThemeUI 0.3 reexports @theme-ui/components. We want to provide typings for these reexported components, so we can have typechecking and autoimports while using theme-ui.

Comment (prateekkathal) [reply types/theme-ui/index.d.ts]: @hasparus Just to be clear, re-exporting can also be understood in terms of making improvements. Also, theme-ui exporting it's own components package directly is completely different than a types package exporting a theme-ui package.

Again, like I said... this package is only meant to define & export types, not the whole library or components. If this is the case, then this PR doesn't make sense at all. The package is meant to export only types not the whole components itself. 

Also, just to be clear, theme-ui doesn't support types locally, so there is no way type checking can happen. We have to manually define them. Removing them altogether defeats the purpose of this package itself.

Comment (prateekkathal) [reply types/theme-ui/index.d.ts]: @hasparus It seems I completely overlooked the fact that a types/theme-ui components was released by you. 

Hmm... I think in that case, my comments are irrelevant! 

Thanks for making the PR 👍

**Comment by prateekkathal** [comment]
@hasparus I am gonna wait on the updated version of Styled System types to be merged and will then approve this one. 

Thanks again for the PR! 👍

File: types/theme-ui/index.d.ts (hunk 1 above)
Comment (hasparus) [reply types/theme-ui/index.d.ts]: Sorry I didn’t express myself clearly. I assumed "types" is implicit in the context of DefinitelyTyped. A bit lazy of me.

**Review by prateekkathal — APPROVED** [review]

**Comment by prateekkathal** [comment]
@hasparus The commit from #41903 requires `Omit`. Maybe we should import it from the components package.

**Pull request merged**